src/v/iceberg/values_avro.cc

```diff
@@ -0,0 +1,349 @@
+namespace iceberg {
+
+namespace {
+
+// Returns the Avro datum corresponding to the given value. The value is
+// expected to match the given Avro schema.
+//
+// The given value may be null if the schema refers to an optional value.
+avro::GenericDatum
+val_to_avro(const std::optional<value>&, const avro::NodePtr&);
+avro::GenericDatum base_val_to_avro(const value&, const avro::NodePtr&);
+
+// XXX: need to figure out how to correctly instantiate avro::GenericFixed to
+// support the unimplemented fields. Or move to an impl that uses iobufs. Throw
+// for now.
+struct primitive_value_avro_visitor {
+    template<typename ValueT>
+    avro::GenericDatum operator()(const ValueT& v) {
+        return {v.val};
+    }
+    avro::GenericDatum operator()(const string_value& v) {
+        const auto size_bytes = v.val.size_bytes();
+        iobuf_const_parser buf(v.val);
+        return {buf.read_string(size_bytes)};
+    }
+    avro::GenericDatum operator()(const binary_value& v) {
+        std::vector<uint8_t> data;
+        auto bytes = iobuf_to_bytes(v.val);
+        data.resize(bytes.size());
+        for (size_t i = 0; i < bytes.size(); i++) {
+            data[i] = bytes[i];
+        }
+        return {data};
+    }
+    avro::GenericDatum operator()(const decimal_value&) {
+        throw std::invalid_argument("XXX decimal not implemented");
+    }
+    avro::GenericDatum operator()(const fixed_value&) {
+        throw std::invalid_argument("XXX fixed not implemented");
+    }
+    avro::GenericDatum operator()(const uuid_value&) {
+        throw std::invalid_argument("XXX uuid not implemented");
+    }
+};
+
+struct value_avro_visitor {
+    explicit value_avro_visitor(const avro::NodePtr& avro_schema)
+      : avro_schema_(avro_schema) {}
+
+    // Expected to match with the value on which the caller is calling
+    // operator().
+    const avro::NodePtr& avro_schema_;
+
+    avro::GenericDatum operator()(const primitive_value& v) {
+        return std::visit(primitive_value_avro_visitor{}, v);
+    }
+    avro::GenericDatum operator()(const std::unique_ptr<struct_value>& v) {
+        return struct_to_avro(*v, avro_schema_);
+    }
+    avro::GenericDatum operator()(const std::unique_ptr<list_value>& v) {
+        if (avro_schema_->leaves() != 1) {
+            throw std::invalid_argument(
+              fmt::format("Expected 1 leaf: {}", avro_schema_->leaves()));
+        }
+        avro::GenericDatum datum(avro_schema_);
+        auto& arr = datum.value<avro::GenericArray>();
+        const auto& child_schema = avro_schema_->leafAt(0);
+        for (const auto& element_ptr : v->elements) {
+            arr.value().emplace_back(val_to_avro(element_ptr, child_schema));
+        }
+        return datum;
+    }
+    avro::GenericDatum operator()(const std::unique_ptr<map_value>& v) {
+        avro::GenericDatum datum(avro_schema_);
+        auto& m = datum.value<avro::GenericArray>();
+        const auto& kv_schema = avro_schema_->leafAt(0);
```

Comment:
in the `list_value` arm you checked that there is exactly one leaf in `avro_schema_`. Would that be appropriate here too?

Reply:
Done